### PR TITLE
Add ready() Promises for TDS storage and extension startup

### PR DESCRIPTION
**Pull request by kzar**
There are numerous race conditions caused by attempting to use
extension configuration before it has finished loaded. These race
conditions can cause exceptions in the background page on extension
install, exceptions in the first run page and integration test
failures.

To resolve some of those let's:
 - Split the extension startup logic out into a separate module and
   provide a ready() Promise that can be checked for when the
   extension's startup sequence has been completed. That is especially
   useful, since it provides an easy way to know that the Trackers
   Object is ready for use.
 - Provide a TDS storage ready() function that can be used to check if
   all TDS configurations (or one specific configuration) have
   finished loading. That is useful, for example, to wait for the
   Click to Load configuration to finish loading before returning it
   to the Click to Load content script.

**Reviewer:** @jonathanKingston 

## Steps to test this PR:
- Ensure integration tests still pass.
- Ensure that there are no exceptions shown in the console for the first run page after the extension is first installed.
- Ensure that there are no Click to Load exceptions in the background console when the extension is first installed.

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/js/background/startup.es6.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+const tdsStorage = require('./storage/tds.es6')
+const trackers = require('./trackers.es6')
+const { fetchAlias, showContextMenuAction } = require('./email-utils.es6')
+
+/** @module */
+
+let resolveReadyPromise
+const readyPromise = new Promise(resolve => { resolveReadyPromise = resolve })
+
+const onStartup = async () => {
```
Comment (jonathanKingston) [inline shared/js/background/startup.es6.js:17]:
```suggestion
async function onStartup () {
```

File: shared/js/background/storage/tds.es6.js [hunk 2]
```diff
@@ -221,12 +252,24 @@ class TDSStorage {
...
     }
+
+    ready (configName) {
+        if (!configName) {
+            return Promise.all(this._onReadyPromises.values())
+        }
+
+        const readyPromise = this._onReadyPromises.get(configName)
+        if (!readyPromise) {
+            throw new Error('Unknown configuration: ' + configName)
```
Comment (jonathanKingston) [inline shared/js/background/storage/tds.es6.js:270]:
```suggestion
            throw new Error(`Unknown configuration: ${configName}`)
```

File: shared/js/background/storage/tds.es6.js [hunk 3]
```diff
@@ -12,12 +12,29 @@ class TDSStorage {
...
+
         this.isInstalling = false
-        this.onUpdatedListeners = new Map()
+
+        this._onUpdatedListeners = new Map()
+
+        this._onReadyResolvers = new Map()
+        this._onReadyPromises = new Map()
+        for (const configName of
+            ['tds', 'surrogates', 'ClickToLoadConfig', 'config']) {
```
Comment (jonathanKingston) [inline shared/js/background/storage/tds.es6.js:28]: I think this can be derived from tdsLists[].name from constants.

Comment (kzar) [reply shared/js/background/storage/tds.es6.js]: Good point! Done

File: shared/js/background/storage/tds.es6.js (hunk 2 above)
Comment (kzar) [reply shared/js/background/storage/tds.es6.js]: Done

File: shared/js/background/startup.es6.js (hunk 1 above)
Comment (kzar) [reply shared/js/background/startup.es6.js]: Done

**Review by jonathanKingston — APPROVED** [review]
LGTM thanks for walking me through it all over video :)

**Pull request merged**